src/backend/impl/src/repositories/proposal_repository.rs

```diff
@@ -30,33 +35,89 @@ impl ProposalRepository for ProposalRepositoryImpl {
         STATE.with_borrow(|s| s.proposals.get(proposal_id))
     }
 
-    fn get_proposals(&self) -> Vec<(ProposalId, Proposal)> {
-        STATE.with_borrow(|s| {
-            s.proposals_sorted_index
-                .iter()
-                .map(|((_, id), _)| (id, s.proposals.get(&id).unwrap()))
+    fn get_proposals(
+        &self,
+        state: Option<ReviewPeriodState>,
+    ) -> Result<Vec<(ProposalId, Proposal)>, ApiError> {
+        let range = ProposalStatusTimestampRange::new(state)?;
+
+        let proposals = STATE.with_borrow(|s| {
+            s.proposals_status_timestamp_index
+                .range(range)
+                .map(|(_, id)| (id, s.proposals.get(&id).unwrap()))
                 .collect()
-        })
+        });
+
+        Ok(proposals)
+    }
+
+    fn complete_pending_proposals(&self, current_time: DateTime) -> Result<(), ApiError> {
+        let range = ProposalStatusTimestampRange::new(ReviewPeriodState::InProgress.into())?;
+
+        STATE.with_borrow_mut(|s| {
+            let pending_proposals: Vec<_> = s
+                .proposals_status_timestamp_index
+                .range(range)
+                .filter_map(|(_, id)| {
+                    s.proposals.get(&id).and_then(|proposal| {
+                        self.is_proposal_pending(&proposal, &current_time)
+                            .then_some((id, proposal))
+                    })
+                })
+                .collect();
+
+            for (proposal_id, proposal) in pending_proposals {
+                let old_key = ProposalStatusTimestampKey::new(
+                    ReviewPeriodState::InProgress.into(),
+                    proposal.proposed_at,
+                    proposal_id,
+                )?;
+                let new_key = ProposalStatusTimestampKey::new(
+                    ReviewPeriodState::Completed.into(),
+                    proposal.proposed_at,
+                    proposal_id,
+                )?;
+
+                s.proposals_status_timestamp_index.remove(&old_key);
+                s.proposals_status_timestamp_index
+                    .insert(new_key, proposal_id);
+
+                s.proposals.insert(
+                    proposal_id,
+                    Proposal {
+                        state: ReviewPeriodState::Completed,
+                        ..proposal
+                    },
+                );
+            }
+
+            Ok(())
+        })?;
+
+        Ok(())
```

Comment:
This can be removed and we can return the `STATE.with_borrow_mut` directly

Reply:
Good catch, done